nomad/node_endpoint.go

```diff
@@ -261,61 +261,75 @@ func (n *Node) Deregister(args *structs.NodeDeregisterRequest, reply *structs.No
 	}
 
 	// Verify the arguments
-	if args.NodeID == "" {
-		return fmt.Errorf("missing node ID for client deregistration")
+	if len(args.NodeIDs) == 0 {
+		return fmt.Errorf("missing node IDs for client deregistration")
 	}
-	// Look for the node
+
+	// Open state handles
 	snap, err := n.srv.fsm.State().Snapshot()
 	if err != nil {
 		return err
 	}
 
 	ws := memdb.NewWatchSet()
-	node, err := snap.NodeByID(ws, args.NodeID)
-	if err != nil {
-		return err
-	}
-	if node == nil {
-		return fmt.Errorf("node not found")
-	}
 
-	// Commit this update via Raft
-	_, index, err := n.srv.raftApply(structs.NodeDeregisterRequestType, args)
-	if err != nil {
-		n.logger.Error("deregister failed", "error", err)
-		return err
-	}
+	for _, nodeID := range args.NodeIDs {
+		// Look for the node
+		node, err := snap.NodeByID(ws, nodeID)
+		if err != nil {
+			return fmt.Errorf("node lookup failed: %s: %v", nodeID, err)
+		}
+		if node == nil {
+			return fmt.Errorf("node not found: %s", nodeID)
+		}
 
-	// Clear the heartbeat timer if any
-	n.srv.clearHeartbeatTimer(args.NodeID)
+		// Clear the heartbeat timer if any
+		n.srv.clearHeartbeatTimer(nodeID)
```

Comment:
There is a potential regression here, previously we ensured the node was removed from raft before clearing the heartbeat timer, and now we clear the heartbeat timer and revoke accessors first.

This means that we may potentially end up in an invalid state whereby we stop monitoring a node, but fail to remove it from raft.

Reply:
great catch, thanks! that makes sense